src/py/reactpy/reactpy/core/state_recovery.py

```diff
@@ -0,0 +1,295 @@
+            pepper=self._pepper,
+            salt=salt,
+            object_to_type_id=self._object_to_type_id,
+            type_id_to_object=self._type_id_to_object,
+            max_object_length=self._max_object_length,
+            max_num_state_objects=self._max_num_state_objects,
+            default_serializer=self._default_serializer,
+            deserializer_map=self._deserializer_map,
+        )
+
+
+class StateRecoverySerializer:
+
+    def __init__(
+        self,
+        totp: pyotp.TOTP,
+        target_time: float | None,
+        otp_max_age: int,
+        otp_mixer: float,
+        pepper: str,
+        salt: str,
+        object_to_type_id: dict[Any, bytes],
+        type_id_to_object: dict[bytes, Any],
+        max_object_length: int,
+        max_num_state_objects: int,
+        default_serializer: Callable[[Any], bytes] | None = None,
+        deserializer_map: dict[type, Callable[[Any], Any]] | None = None,
+    ) -> None:
+        self._totp = totp
+        self._otp_mixer = otp_mixer
+        target_time = target_time or time.time()
+        self._target_time = target_time
+        otp_code = self._get_otp_code(target_time)
+        self._otp_max_age = otp_max_age
+        self._otp_code = otp_code.encode("utf-8")
+        self._pepper = pepper.encode("utf-8")
+        self._salt = salt.encode("utf-8")
+        self._object_to_type_id = object_to_type_id
+        self._type_id_to_object = type_id_to_object
+        self._max_object_length = max_object_length
+        self._max_num_state_objects = max_num_state_objects
+        self._provided_default_serializer = default_serializer
+        deserialization_map = {
+            datetime.timezone: lambda x: datetime.timezone(
+                datetime.timedelta(**x["offset"]), x["name"]
+            ),
+        }
+        self._deserializer_map = deserialization_map | (deserializer_map or {})
+
+    def _get_otp_code(self, target_time: float) -> str:
+        at = self._totp.at
+        return f"{at(target_time)}{at(target_time - self._otp_mixer)}{at(target_time + self._otp_mixer)}"
+
+    async def serialize_state_vars(
+        self, state_vars: dict[str, Any]
+    ) -> dict[str, tuple[str, str, str]]:
+        if len(state_vars) > self._max_num_state_objects:
+            logger.warning(
+                f"State is too large ({len(state_vars)}). State will not be sent"
+            )
+            return {}
+        result = {}
+        for chunk in chunked(state_vars.items(), 50):
+            for key, value in chunk:
+                result[key] = self._serialize(key, value)
+            await asyncio.sleep(0)  # relinquish CPU
+        return result
+
+    def _serialize(self, key: str, obj: object) -> tuple[str, str, str]:
+        type_id = b"1"  # bool
+        if obj is None:
+            return "0", "", ""
+        match obj:
+            case True:
+                result = b"true"
+            case False:
+                result = b"false"
```

Comment:
This extra complexity is for performance reasons.

`None` is not signed.
Boolean values are really common so this is a shortcut.